internal/api_client/task_tracker.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 Broadcom. All Rights Reserved.
+// SPDX-License-Identifier: MPL-2.0
+
+package api_client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/vmware/terraform-provider-vcf/internal/constants"
+	"github.com/vmware/vcf-sdk-go/client"
+	"github.com/vmware/vcf-sdk-go/client/tasks"
+	"github.com/vmware/vcf-sdk-go/models"
+	"time"
+)
+
+const (
+	defaultPollingInterval = time.Second * 20
+)
+
+type TaskTracker struct {
+	ctx             context.Context
+	client          *client.VcfClient
+	taskId          string
+	pollingInterval time.Duration
+
+	completedTasks map[string]bool
+}
+
+func NewTaskTracker(ctx context.Context, client *client.VcfClient, taskId string) *TaskTracker {
+	return &TaskTracker{
+		ctx:             ctx,
+		client:          client,
+		taskId:          taskId,
+		pollingInterval: defaultPollingInterval,
+		completedTasks:  make(map[string]bool),
+	}
+}
+
+func NewTaskTrackerWithCustomPollingInterval(ctx context.Context, client *client.VcfClient, taskId string, pollingInterval time.Duration) *TaskTracker {
+	tracker := NewTaskTracker(ctx, client, taskId)
+	tracker.pollingInterval = pollingInterval
+	return tracker
+}
+
+func (t *TaskTracker) WaitForTask() error {
+	ticker := time.NewTicker(t.pollingInterval)
+
+	for {
+		select {
+		case <-t.ctx.Done():
+		case <-ticker.C:
+			task, err := t.getTask()
+			if err != nil {
+				return err
+			}
+
+			t.logTask(task)
+
+			if task.Status == "In Progress" || task.Status == "Pending" || task.Status == "IN_PROGRESS" {
+				return nil
+			}
+
+			if task.Status == "Failed" || task.Status == "Cancelled" {
+				errorMsg := fmt.Sprintf("Task with ID = %s , Name: %q Type: %q is in state %s",
+					task.ID, task.Name, task.Type, task.Status)
+				tflog.Error(t.ctx, errorMsg)
+
+				return errors.New(errorMsg)
+			}
+
+			tflog.Info(t.ctx, fmt.Sprintf("Task with ID = %s is in state %s, completed at %s",
+				task.ID, task.Status, task.CompletionTimestamp))
+			return nil
+		}
+	}
```

Comment:
You could then use the constants and a switch:

```suggestion
    for {
        select {
        case <-t.ctx.Done():
            return t.ctx.Err()
        case <-ticker.C:
            task, err := t.getTask()
            if err != nil {
                return err
            }

            t.logTask(task)

            switch task.Status {
            case statusInProgress, statusPending:
                continue
            case statusFailed, statusCancelled:
                errorMsg := fmt.Sprintf("Task with ID = %s, Name: %q, Type: %q is in state %s",
                    task.ID, task.Name, task.Type, task.Status)
                tflog.Error(t.ctx, errorMsg)
                return errors.New(errorMsg)
            default:
                tflog.Info(t.ctx, fmt.Sprintf("Task with ID = %s is in state %s, completed at %s",
                    task.ID, task.Status, task.CompletionTimestamp))
                return nil
            }
        }
    }
```